File: FollowingInteger/fi.py
#following integer soln in py for code eval by steven a dunn

import sys

def find_largest(digits):
    largest_digit = digits[0]
    index = 0

    for i in range(len(digits)):
        if digits[i] >= largest_digit:
            largest_digit = digits[i]
            index = i

    return index

def percolate(digits, largest_index):
    if largest_index > 0:
        digits[largest_index], digits[largest_index-1] = digits[largest_index-1], digits[largest_index]
    else:
        #count the zeros
        #remove zeros
        #sort
        #add zero count + 1 to the num
        zero_count = 0

        print "need to expand number"
    return int(''.join(digits))



f = open(sys.argv[1], 'r')

for line in f:
    digits = list(line)
    largest_index = find_largest(digits)
    num = percolate(digits, largest_index)
    print num
f.close()

File: FollowingInteger/cpp/FI.cpp
// Following Integer solution in C++ for CodeEval.com by Steven A. Dunn

#include <algorithm>
#include <cstdio>
#include <cstdlib>
#include <fstream>
#include <string>

using std::ifstream;
using std::next_permutation;
using std::sort;
using std::string;

string parse_digits(string);

int main(int argc, char* argv[])
{
  ifstream inputFile(argv[1]);
  string line;

  if(inputFile)
  {
    while(getline(inputFile, line))
    {
      int original = atoi(line.c_str());
      string digits = parse_digits(line);
      sort(digits.begin(), digits.end());
      while (next_permutation(digits.begin(), digits.end()))
      {
        int val = atoi(digits.c_str());
        if (val > original)
        {
          printf("%d\n", val);
          break;
        }
      }
    }
    inputFile.close();
  }
  return 0;
}

string parse_digits(string line)
{
  sort(line.begin(), line.end());
  int remaining = 6 - line.size();
  string padding;
  for (int i = 0; i < remaining; ++i)
    padding += "0";
  return (padding + line);
}